### DISCUSSION TITLE
Custom compiler extension for WiX 4 bundle

### DISCUSSION THREAD

**Discussion by yuvnith** [wixtoolset/issues#7788] [Questions] [answered]
We have a WiX installer which was developed using WiX 3, now we are upgrading it to use WiX 4. 
We use our own Managed Bootstrapper Application and have our custom bundle elements, to compile these custom elements we were using a Skeleton WiX Extension in WiX 3 and use that as a build parameter.

-ext "C:\Repo\XYZ.Compiler.dll"

We took [https://github.com/wixtoolset/wix/tree/develop/src/ext](url) as a reference and modified our custom compiler extension.
**This is how we implemented the BaseExtensionFactory**
public class ExtensionFactory : BaseExtensionFactory
    {
        protected override IReadOnlyCollection<Type> ExtensionTypes => new[]
         {
            typeof(XYZExtensionData),
            typeof(XYZCompiler)
        };
    }
 
When we build our project, we are seeing the below error.
WIX0125	The extension 'XYZ.Compiler' uses the same xml schema namespace, 'XYZNamespace', as previously loaded extension 'XYZ.Compiler'. Please either remove one of the extensions or rename the xml schema namespace to avoid the collision.

Can anyone please help us in finding the root cause.
Thanks in advance.

**Comment by robmen** [comment]
It sounds like you're resuing your XML namespace across extensions.

**Reply by yuvnith** [reply]
Hi, we only have one extension and that is a compiler extension. The same extension is mentioned twice in the error. Seems like it is getting loaded multiple times.

**Reply by robmen** [reply]
Interesting hypothesis. Might be a place to start your debugging journey.

**Reply by yuvnith** [reply]
**Update**: I looked into the MSBUILD Temp file(.rsp), which is sent to WIX as an argument. Observed that the build command has -ext "mycompilerextension.dll" argument twice, even though my csproj has a single argument in **CompilerAdditionalOptions** section

```
<PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x86'">
    <CompilerAdditionalOptions>-ext "mycompilerextension.dll"</CompilerAdditionalOptions>
  </PropertyGroup>

```
I need to understand this visual studios build/MSBuild behavior  

Note: I have run the wix build command available in the .rsp file, without including the duplicate extension argument, and it works fine.

**Comment by yuvnith** [comment, marked as answer]
[WixToolset.BuildTasks.WixBuild](https://github.com/wixtoolset/wix/tree/b84202794f5fd4f21e0fb1fc74c2f5d531035db3/src/wix/WixToolset.BuildTasks/WixBuild.cs) adds the Additional Options(line 85). There is also a call to the [base method](https://github.com/wixtoolset/wix/tree/develop/src/internal/WixInternal.BaseBuildTasks.Sources/BaseToolsetTask.cs) which also adds the Additional Options(line 92). This explains the issue I am facing. 

@robmen Is this an issue ?

**Reply by robmen** [reply]
Yep, that looks like a bug.

**Reply by yuvnith** [reply]
Thank you for the confirmation.

**Reply by yuvnith** [reply]
Hi @robmen, from this conversation, I have created a [bug](https://github.com/wixtoolset/issues/issues/7837).
